Add generic OIDC authentication flow

palash247 opened the pull request:
## What type of PR is this? 
<!-- Check all that apply, delete what doesn't apply. -->

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] New Query Runner (Data Source) 
- [ ] New Alert Destination
- [ ] Other

## Description
<!-- In case of adding / modifying a query runner, please specify which version(s) you expect are compatible. -->
### Generic OIDC login

_<ins>Note: Seeking feedback</ins>_

### Problem

Currently, Redash is limited to supporting authentication via Google's OIDC. Expanding the authentication support to a generic OIDC flow increases Redash's accessibility and flexibility, making it a more inclusive tool for diverse environments.

### Implementation Steps

To fully implement and integrate the generic OIDC authentication flow into Redash, the following steps need to be taken (Open for feedback):

- [x] **Make OIDC Discovery Endpoint Configurable**: This flexibility allows Redash to dynamically adapt to different IdPs by simply modifying configuration parameters.

- [x] **Implement Routes for OIDC Authentication**: Develop new routes within the application to handle the authentication process via OIDC.

- [x] **Modify Login Templates**: Update the UI components, specifically the login page templates, to include an OIDC login button.

- [ ] **Manual Testing with Multiple OIDC IdPs**: To ensure compatibility and robustness, manually test the new authentication flow with multiple OIDC IdPs, including but not limited to Google, AWS, Okta, and Auth0. (Tested with Google and AWS Cognito)

- [x] **Implement Unit Tests**

- [ ] **Write Documentation for OIDC Configuration**: Create detailed documentation to assist users in configuring OIDC with Redash. This documentation should cover configuration steps for different IdPs.

- [x] **Mutate OIDC to include domain check**: (not sure about this one) Current Google OIDC flow has domain verification, see if this can be made generic.

- [ ] **Replace Google OIDC with generic one**

## How is this tested?

- [x] Unit tests (pytest, jest)
- [ ] E2E Tests (Cypress)
- [x] Manually
- [ ] N/A

<!-- If Manually, please describe. -->
For testing manually, following steps should be more are less common among different IdPs.

1. Configure an IdP with OIDC federation protocol and get the discovery endpoint, client id and secret id.
2. Set 3 environment variables accordingly:
 - REDASH_OIDC_CLIENT_ID
 - REDASH_OIDC_CLIENT_SECRET
 - REDASH_OIDC_ISSUER_URL
 3. Check if Login with OIDC button available in the login page.
 4. Login
 
 - [x] AWS Cognito
 - [x] Google
 - [x] Okta
 - [ ] Auth0
 - [ ] Azure


## Related Tickets & Documents
<!-- If applicable, please include a link to your documentation PR against getredash/website -->
#6781 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
<img width="524" alt="Screenshot 2024-02-27 at 2 43 36 PM" src="https://github.com/getredash/redash/assets/10260127/89036a80-231b-4923-9152-e08d7f3aaaff">
<img width="969" alt="Screenshot 2025-02-18 at 11 40 13 PM" src="https://github.com/user-attachments/assets/2dafbc7a-a6ce-42d8-9f8c-61627716ed9e" />

palash247 committed:
Add generic OIDC authentication flow (#6781)

- Added a generic authentication flow for OIDC.
- Modified some templates to display OIDC login button.
- Added new env variables to support the OIDC auth flow. Mainly, OIDC
  discovery endpoint is configurable now.
- Added new routs for OIDC auth.
- Manually tested same flow with Google (without domain verification) and AWS Cognito.

lucasfcnunes commented on redash/settings/__init__.py:
```diff
@@ -138,6 +138,11 @@
 
+OIDC_CLIENT_ID = os.environ.get("REDASH_OIDC_CLIENT_ID", "")
+OIDC_CLIENT_SECRET = os.environ.get("REDASH_OIDC_CLIENT_SECRET", "")
+OIDC_COFIGURATION_URL = os.environ.get("REDASH_OIDC_COFIGURATION_URL", "")
```
`CO(N)FIGURATION` spelled wrongly

Usually we see `*ISSUER_URL` or `*WELL_KNOWN_URL` as the env var names to the `$1/.well-known/openid-configuration` json to get the other endpoints.

justinclift replied:
> CO(N)FIGURATION spelled wrongly

Oops.  We should probably try to fix that somehow (without causing issues for our users). :smile:

palash247 replied:
Oops, let me fix that.
Also, I need some feedback on the overall approach. I am thinking of moving all OIDC logins through this code, including Google's. To keep things backward compatible, will keep Google's env variables and map them to new generic ones. The domain verification logic will only work for Google OIDC in the generic flow.
Thoughts?

palash247 replied:
Oops! Let me fix that.

palash247 committed:
Add unit tests for generic oidc

  - Fixed typo
  - Added unit tests cases

palash247 commented:
Hi @arikfr,
Can I please get some feedback over the approach?

lucasfcnunes commented on redash/authentication/oidc.py:
```diff
@@ -0,0 +1,106 @@
+        else:
+            org = current_org
+
+        user = create_and_login_user(org, profile["name"], profile["email"])
```
1. maybe allow choosing what field to use as the `name`?
2. if email not verified, should we allow user creation? (avoid possible future CVE)

palash247 replied:
1. For now I am thinking of creating name based on what claims are received from the scope profile.
2. Aligned, adding a check for claim `email_verified`.

palash247 commented on redash/authentication/oidc.py:
```diff
@@ -0,0 +1,106 @@
+        headers = {"Authorization": "Bearer {}".format(access_token)}
+        response = requests.get(oidc_config["userinfo_endpoint"], headers=headers)
+
+        if response.status_code == 401:
+            logger.warning("Failed getting user profile (response code 401).")
+            return None
```
Sorry for responding so late.

I think checking for 401 is important. But I get your point, it should also handle other non 200 HTTP status code.

palash247 commented on redash/authentication/oidc.py:
```diff
@@ -0,0 +1,106 @@
+        name="oidc",
+        server_metadata_url=settings.OIDC_ISSUER_URL,
+        client_kwargs={
+            "scope": "openid email profile",
```
Agreed. Making it configurable.

palash247 replied:
I am thinking of making scope configurable with default being: "openid email profile". If something different is configured, we'll still make sure email and profile are included. Scope `email` for claims `email` and  `email_verified`. Scope `profile` for the name and picture (if available).

palash247 committed:
have option to setup valid domains to login from

- removed redudent userinfo fetch instead parsed id_token for userinfo
- added domains verification for OIDC in UI
- incorporated review comments
- tested with AWS cognito

palash247 committed:
add unit tests for oidc auth flow

- fix typo

lucasfcnunes commented on redash/authentication/oidc.py:
```diff
@@ -0,0 +1,142 @@
+        logger.debug("Authorized user inbound")
+
+        token = oauth.oidc.authorize_access_token()
+        user_info = oauth.oidc.parse_id_token(token)
+        if user_info:
+            session["user"] = user_info
```
I would `user_info` -> `userinfo` in every single place/file because the endpoint is [`/userinfo`](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoRequest) haha (non-issue really)

lucasfcnunes reviewed (dismissed):
LGTM! Let's see how the tests do!

Thanks for your work @palash247. I will approve this and leave polishing for another PR. This is a very needed feature and it can't wait anymore.

lucasfcnunes commented:
@palash247 

Can you check:
- https://github.com/getredash/redash/actions/runs/13403828487/job/37655814191?pr=6783 (formating)
- https://github.com/getredash/redash/actions/runs/13403828487/job/37655814191?pr=6783 (oidc tests failing)

*You can force push to dismiss the version bumps as you did before

lucasfcnunes commented:
@palash247, can I continue the work (in another branch) from https://github.com/getredash/redash/pull/6783/commits/f9f5c9b80d1c203e5b67d829260dc1610927eac3?

palash247 committed:
fix OIDC auth issues from PR review feedback

- Remove duplicate OAuth(app) initialization
- Remove unused access_token validation block
- Fix has_user() boolean comparison (== 1 -> truthy check)
- Change domain logging from INFO to DEBUG level
- Add show_oidc_login to <hr> separator in login and invite templates
- Fix test: use oidc.create_and_login_user instead of google oauth version
- Fix test: remove erroneous url_prefix causing duplicate org_slug parameter

palash247 committed:
improve OIDC auth: add discovery URL builder, error handling, and explicit client credentials

- Add _build_server_metadata_url() to accept both issuer URL and full .well-known URL
- Add explicit client_id and client_secret to oauth.register() call
- Add try/except around authorize_access_token() for proper error handling
- Remove unused session["user"] assignment
- Add unit tests for _build_server_metadata_url()

palash247 committed:
guard parse_id_token against JWT decoding/validation exceptions

palash247 committed:
narrow parse_id_token exception to AuthlibBaseError

Commit:
address copilot review feedback for OIDC auth

- add OIDC_SCHEME_OVERRIDE setting for reverse proxy compatibility
- pass org_slug in error redirect paths for MULTI_ORG support
- handle missing email claim from OIDC provider gracefully
- treat missing email_verified claim as verified (optional per OIDC spec)
- add alt="" to OIDC login button images for accessibility
- clean up whitespace in OIDCLoginSettings.jsx alert message
- fix typo: test_creates_vaild_new_user -> test_creates_valid_new_user

lucasfcnunes commented:
@justinclift, can you check this out too?

restyled-commits committed:
Restyled by prettier

justinclift commented:
@lucasfcnunes Sorry, but I don't put time into Redash any more.

palash247 committed:
Merge branch 'master' into master

github-actions[bot] committed:
Snapshot: 26.06.0-dev

kingspride commented:
@eran132 Hi, can you check if this Issue is stuck since @justinclift does not work on redash anymore but is marked as reviewer here?

Thanks!

eran132 commented:
@
@kingspride you're right — this is blocked. @justinclift confirmed above (May 1) he's no longer active on Redash, and he's the only requested reviewer, so the review can't progress as-is. I'm a contributor here, not a maintainer, so I can't reassign reviewers myself.

@arikfr — could you re-triage #6783 and reassign review to an active maintainer? It's `MERGEABLE` but currently `BEHIND master` (so @palash247 would need a quick rebase), and the Tests workflow still needs a maintainer's "approve and run".
@

github-actions[bot] committed:
Snapshot: 26.07.0-dev

github-actions[bot] committed:
Snapshot: 26.08.0-dev
